user_tools/src/as_pytools/storagelib/csppath.py

```diff
@@ -0,0 +1,328 @@
+# limitations under the License.
+
+"""
+Abstract representation of a file path that can access local/URI values.
+Similar to cloudpathlib project, this implementation uses dict registry to
+register an implementation. However, the path representation is built on top of
+pyArrow FS API. As a result, there is no need to write a full storage client to
+access remote files. This comes with a tradeoff in providing limited set of file
+operations.
+"""
+
+import abc
+import sys
+from collections import defaultdict
+from functools import cached_property
+from pathlib import Path as PathlibPath
+from typing import Union, Type, TypeVar, Any, Dict, Callable, overload, Optional, TYPE_CHECKING, List
+
+from pyarrow import fs
+from pyarrow.fs import FileType, FileSystem, FileInfo
+from pydantic import ValidationError, model_validator, FilePath, AnyHttpUrl, StringConstraints, TypeAdapter
+from pydantic.dataclasses import dataclass
+from pydantic_core import PydanticCustomError
+from typing_extensions import Annotated
+
+from ..exceptions import (
+    InvalidProtocolPrefixError,
+    FSMismatchError, CspFileExistsError
+)
+
+from ..utils.util import get_path_as_uri, is_http_file
+
+if sys.version_info >= (3, 10):
+    from typing import TypeGuard
+else:
+    from typing_extensions import TypeGuard
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    from typing_extensions import Self
+
+if TYPE_CHECKING:
+    from .cspfs import CspFs
+
+
+CspPathString = Annotated[str, StringConstraints(pattern=r'^\w+://.*')]
+
+
+class CspPathImplementation:
+    """
+    A metaclass implementation that describes the behavior of the path class
+    """
+    name: str
+    _path_class: Type['CspPath']
+    _fs_class: Type['CspFs']
+    _fslib_class: Type['FileSystem']
+
+    @property
+    def fs_class(self) -> Type['CspFs']:
+        return self._fs_class
+
+    @fs_class.setter
+    def fs_class(self, clazz):
+        self._fs_class = clazz
+
+    @property
+    def path_class(self) -> Type['CspPath']:
+        return self._path_class
+
+    @path_class.setter
+    def path_class(self, clazz):
+        self._path_class = clazz
+
+    @property
+    def fslib_class(self) -> Type['FileSystem']:
+        return self._fslib_class
+
+    @fslib_class.setter
+    def fslib_class(self, clazz):
+        self._fslib_class = clazz
+
+
+path_impl_registry: Dict[str, CspPathImplementation] = defaultdict(CspPathImplementation)
+
+T = TypeVar('T')
+CspPathT = TypeVar('CspPathT', bound='CspPath')
+
+
+def register_path_class(key: str) -> Callable[[Type[CspPathT]], Type[CspPathT]]:
+    def decorator(cls: Type[CspPathT]) -> Type[CspPathT]:
+        if not issubclass(cls, CspPath):
+            raise TypeError('Only subclasses of CloudPath can be registered.')
+        path_impl_registry[key].path_class = cls
+        cls._path_meta = path_impl_registry[key]  # pylint: disable=protected-access
+        return cls
+
+    return decorator
+
+
+@dataclass
```

Comment:
Curios why this is from pydantic.dataclasses instead of just dataclasses

Reply:
Nevermind. I think it's required because of the model_validator which is used in this class.